exporter/signalfxexporter/metricdata_to_signalfxv2.go

```diff
@@ -0,0 +1,403 @@
+		&count,
+		&sum)
+
+	// 3. Each quantile value is converted to a gauge called <basename>_quantile
+	// and will include a dimension called quantile that specifies the quantile.
+	percentiles := summaryValue.GetSnapshot().GetPercentileValues()
+	if percentiles == nil {
+		return sfxDataPoints, fmt.Errorf(
+			"unknown percentiles values for summary metric %q",
+			*sfxBaseDataPoint.Metric)
+	}
+	metricName := *sfxBaseDataPoint.Metric + "_quantile"
+	for _, quantile := range percentiles {
+
+		// Adding the "quantile" dimension.
+		quantileDimensions := make([]*sfxpb.Dimension, len(sfxBaseDataPoint.Dimensions)+1)
+		copy(quantileDimensions, sfxBaseDataPoint.Dimensions)
+
+		// TODO: What happens if a dimension "quantile" was already specified?
+		//  symmetrical for distribution, and for others too.
+		quantileDimensions[len(quantileDimensions)-1] = &sfxpb.Dimension{
+			Key:   &quantileDimensionKey,
+			Value: float64ToDimValue(quantile.Percentile),
+		}
+
+		quantileDP := *sfxBaseDataPoint
+		quantileDP.Dimensions = quantileDimensions
+		quantileDP.Metric = &metricName
+		quantileDP.MetricType = quantileMetricType
+		value := quantile.Value
+		quantileDP.Value = &sfxpb.Datum{DoubleValue: &value}
+
+		sfxDataPoints = append(sfxDataPoints, &quantileDP)
+	}
+
+	return sfxDataPoints, nil
+}
+
+func appendTotalAndSum(
+	sfxDataPoints []*sfxpb.DataPoint,
+	sfxBaseDataPoint *sfxpb.DataPoint,
+	count *int64,
+	sum *float64,
+) []*sfxpb.DataPoint {
+
+	sfxDataPoints = append(
+		sfxDataPoints,
+		buildTotalDataPoint(sfxBaseDataPoint, count))
+
+	sfxDataPoints = append(
+		sfxDataPoints,
+		buildSumDataPoint(sfxBaseDataPoint, sum))
+
+	return sfxDataPoints
+}
+
+func buildTotalDataPoint(
+	sfxBaseDataPoint *sfxpb.DataPoint,
+	count *int64,
+) *sfxpb.DataPoint {
+
+	totalCountDP := *sfxBaseDataPoint
+	totalCountName := *sfxBaseDataPoint.Metric + "_count"
+	totalCountDP.Metric = &totalCountName
+	totalCountDP.MetricType = totalCountMetricType
+	totalCountDP.Value = &sfxpb.Datum{IntValue: count}
+
+	return &totalCountDP
+}
+
+func buildSumDataPoint(
+	sfxBaseDataPoint *sfxpb.DataPoint,
+	sum *float64,
+) *sfxpb.DataPoint {
+
+	sumDP := *sfxBaseDataPoint
+	sumDP.MetricType = sumMetricType
+	sumDP.Value = &sfxpb.Datum{DoubleValue: sum}
+
+	return &sumDP
+}
+
+func filterKeyChars(str string) string {
+	filterMap := func(r rune) rune {
+		if unicode.IsLetter(r) || unicode.IsDigit(r) || r == '_' || r == '-' {
+			return r
+		}
+		return '_'
+	}
+
+	return strings.Map(filterMap, str)
+}
+
+func float64ToDimValue(f float64) *string {
+	// Parameters below are the same used by Prometheus
+	// see https://github.com/prometheus/common/blob/b5fe7d854c42dc7842e48d1ca58f60feae09d77b/expfmt/text_create.go#L450
+	// SignalFx agent uses a different pattern
+	// https://github.com/signalfx/signalfx-agent/blob/5779a3de0c9861fa07316fd11b3c4ff38c0d78f0/internal/monitors/prometheusexporter/conversion.go#L77
+	// TODO: Create a config option and allow user to choose it?
```

Comment:
The way you have it is probably better and it doesn't matter too much as long as it is consistent within the same agent.